Implement handling RFCs

xno-miner opened the pull request:
Resolves https://github.com/ubiquity/work.ubq.fi/issues/54

Changes:
1. Added the `DEVPOOL_RFC_OWNER_NAME` and `DEVPOOL_RFC_REPO_NAME` variables which point to the repository where RFC issues will be added. It's set to `ubiquity/devpool-rfc` which doesn't exist yet.
2. In `createDevPoolIssue()` the issue is added to `devpool-directory` if it has a price label and to `devpool-rfc` if it doesn't
3. In `index.ts`, each issue is searched for in `devpool-directory` as well as in `devpool-rfc`.
4. `applyStateChanges()` creates new variables: `isInRfcRepo`, and `hasCorrectPriceLabels`. These variables are now used for creating the state changes:
    - `noPriceLabels_Close` behavior is unchanged, but it's adjusted to use the variables
    - new `RFCPriceLabels_Close` closes an RFC if it has a price label
    - `issueReopenedMerged_Open` and `issueUnassigned_Open` were adjusted to correctly reopen RFCs


On a side note, while writing this description i have noticed that the capitalization of DevPool/Devpool is inconsistent in the code. For example there's `getDevpoolIssueLabels()` and `handleDevPoolIssue()`. I think it should be standardized to avoid future problems.

xno-miner committed:
Update github.ts

xno-miner committed:
Update index.ts

xno-miner committed:
Minor fix

0x4007 commented on helpers/github.ts:
```diff
@@ -500,10 +505,16 @@ async function applyStateChanges(projectIssues: GitHubIssue[], projectIssue: Git
       comment: "Closed (no price labels)",
     },
+    // HAS price labels set and open in the RFC devpool
+    RFCPriceLabels_Close: {
```
```suggestion
    rfcPriceLabelsClosed: {
```

xno-miner committed:
Update helpers/github.ts

Co-authored-by: アレクサンダー.eth <4975670+0x4007@users.noreply.github.com>

xno-miner commented:
seems like the tests are using `handleDevPoolIssue()` without the new parameter. Since the tests are made for non-RFCs they should just pass false everywhere. I'll add that in a sec

xno-miner committed:
Update devpool-issue-handler.test.ts

xno-miner committed:
Update devpool-issue-handler.test.ts

xno-miner committed:
Update index.ts

xno-miner committed:
Update github.ts

xno-miner commented:
To avoid more potential issues, i changed back all function parameters to original and moved the creation of `isRFC` and `hasCorrectPriceLabels` to `applyStateChanges()`

xno-miner committed:
Minor fix

xno-miner committed:
Minor clean

xno-miner committed:
Fix getting repo name

rndquu commented:
@xno-miner What issue this PR solves? I remember there was a discussion but can't find a related github issue.

xno-miner commented:
> @xno-miner What issue this PR solves? I remember there was a discussion but can't find a related github issue.

https://github.com/ubiquity/work.ubq.fi/issues/54

0x4007 commented:
> > @xno-miner What issue this PR solves? I remember there was a discussion but can't find a related github issue.
> 
> [ubiquity/work.ubq.fi#54](https://github.com/ubiquity/work.ubq.fi/issues/54)

Your pull request body should be:

Resolves https://github.com/ubiquity/work.ubq.fi/issues/54

Read more at https://dao.ubq.fi/devpool-flow#28b52b32bc434aac8d8e6de501ee75e9

rndquu commented on helpers/github.ts:
```diff
@@ -25,6 +25,8 @@ export const projects = _projects as {
 export const DEVPOOL_OWNER_NAME = "ubiquity";
 export const DEVPOOL_REPO_NAME = "devpool-directory";
+export const DEVPOOL_RFC_OWNER_NAME = "ubiquity";
+export const DEVPOOL_RFC_REPO_NAME = "devpool-rfc";
```
Perhaps it makes sense to move `DEVPOOL_OWNER_NAME` and `DEVPOOL_REPO_NAME` to env variables in order not to duplicate variables

xno-miner replied:
moved all of them to env if that's ok

rndquu commented on helpers/github.ts:
```diff
@@ -389,14 +391,14 @@ export async function createDevPoolIssue(projectIssue: GitHubIssue, projectUrl:
-  // if issue doesn't have the "Price" label then skip it, no need to pollute repo with draft issues
-  if (!(projectIssue.labels as GitHubLabel[]).some((label) => label.name.includes(LABELS.PRICE))) return;
+  // if the issue has no price labels, it will be added to RFC repo
+  const hasNoPriceLabels = !(projectIssue.labels as GitHubLabel[]).some((label) => label.name.includes(LABELS.PRICE));
```
With this approach we will [post tweets](https://github.com/ubiquity/devpool-directory/blob/20d74be226e8cae755e44d70d303d5d84acc49c8/helpers/github.ts#L416) for RFC kind issues which should not be the case as far as I understand

xno-miner replied:
fixed

rndquu commented on index.ts:
```diff
@@ -63,13 +66,15 @@ async function main() {
     for (const projectIssue of projectIssues) {
       // if issue exists in devpool
       const devpoolIssue = getIssueByLabel(devpoolIssues, `id: ${projectIssue.node_id}`);
```
Even in a forked repo this code will fetch devpool issues which we don't really need

xno-miner replied:
Could you elaborate please? I don't think i understand

rndquu replied:
My mistake, I actually thought that (as described in the original issue) this repo should be forked and run against `ubiquity/devpool-rfc` hence I [asked](https://github.com/ubiquity/devpool-directory/pull/1212#discussion_r1658756763) for using env variables and [this](https://github.com/ubiquity/devpool-directory/pull/1212#discussion_r1658775169) change.

rndquu commented:
@xno-miner Hey, sorry for a long review, pls take a look at the review comments

xno-miner committed:
Moved the DEVPOOL_ variables to .env (copy the values from .env.example)

xno-miner committed:
Moved DEVPOOL_OWNER_NAME, DEVPOOL_REPO_NAME, DEVPOOL_RFC_OWNER_NAME and DEVPOOL_RFC_REPO_NAME to .env (copy the values from .env.example)

xno-miner committed:
don't tweet rfcs

xno-miner committed:
don't tweet rfcs

rndquu commented:
@xno-miner Combining both devpool and RFC issue logic complicates the code because of newly introduced `if` statements ([one](https://github.com/ubiquity/devpool-directory/pull/1212/files#diff-7f46e46a8d914909f1e349297b52bd5ecad7f9c4d15c83ac06f75f8981d9ebc5R410), [two](https://github.com/ubiquity/devpool-directory/pull/1212/files#diff-7f46e46a8d914909f1e349297b52bd5ecad7f9c4d15c83ac06f75f8981d9ebc5R506), [three](https://github.com/ubiquity/devpool-directory/pull/1212/files#diff-dcdc3e0b3362edb8fec2a51d3fa51f8fb8af8f70247e06d9887fa934834c9122R73)).

Perhaps it's more cleaner to:
1. Create a new branch like `ubiquity/devpool-directory#rfc`
2. Refactor the logic to use only `ubiquity/devpool-rfc` (this eliminates unnecessary `if` statements for distinguishing devpool and RFC issues)
3. Create a new repo `ubiquity/devpool-rfc` and [run](https://github.com/ubiquity/devpool-directory/blob/797536f9bfbb6501756b128bcc4251e1d2e3e061/.github/workflows/sync-issues.yml) `ubiquity/devpool-directory#rfc` there

Right now this is really hard to read.

xno-miner commented:
> @xno-miner Combining both devpool and RFC issue logic complicates the code because of newly introduced `if` statements ([one](https://github.com/ubiquity/devpool-directory/pull/1212/files#diff-7f46e46a8d914909f1e349297b52bd5ecad7f9c4d15c83ac06f75f8981d9ebc5R410), [two](https://github.com/ubiquity/devpool-directory/pull/1212/files#diff-7f46e46a8d914909f1e349297b52bd5ecad7f9c4d15c83ac06f75f8981d9ebc5R506), [three](https://github.com/ubiquity/devpool-directory/pull/1212/files#diff-dcdc3e0b3362edb8fec2a51d3fa51f8fb8af8f70247e06d9887fa934834c9122R73)).
> 
> Perhaps it's more cleaner to:
> 
>     1. Create a new branch like `ubiquity/devpool-directory#rfc`
> 
>     2. Refactor the logic to use only `ubiquity/devpool-rfc` (this eliminates unnecessary `if` statements for distinguishing devpool and RFC issues)
> 
>     3. Create a new repo `ubiquity/devpool-rfc` and [run](https://github.com/ubiquity/devpool-directory/blob/797536f9bfbb6501756b128bcc4251e1d2e3e061/.github/workflows/sync-issues.yml) `ubiquity/devpool-directory#rfc` there
> 
> 
> Right now this is really hard to read.

I think a few `if` statements are cleaner than having two versions of the entire code with only minor differences. I will make a few changes to improve the code readability though.

xno-miner committed:
Moved the DEVPOOL_ variables back from env

xno-miner committed:
Improved code readability

xno-miner committed:
Minor fix

xno-miner committed:
Further improved readability

xno-miner committed:
Minor fix

xno-miner commented:
I removed the inverse logic involving `hasNoPriceLabels` and i think the code is much cleaner now. I also removed the `devpoolRFCIssues` variable and just made `devpoolIssues` contain issues from both repositories.

rndquu commented:
@xno-miner I ran this script 2 times and immediately got rate limited by github API. As far as I understand it happens because now we get issues from 2 repositories [here](https://github.com/ubiquity/devpool-directory/blob/5263f1752c5075db2057f5a46f9634c4db07d68e/index.ts#L40-L41) and overall perform 2x more API requests to github API.

Perhaps it does make sense to:
1. Move RFC parsing logic to a separate branch
2. Move `DEVPOOL_OWNER_NAME` and `DEVPOOL_REPO_NAME` to env variables so that we could use them in a fork
3. Create a new github app (like `ubiquibot-rfc`) so that we could use its token [here](https://github.com/ubiquity/devpool-directory/blob/7d187e44df3378cd7244b90ee01ebe11851e0f64/.github/workflows/sync-issues.yml#L35-L40)
4. Run `ubiquity/devpool-rfc` against this newly created branch

This way we use 2 github apps (1st for `ubiquity/devpool-directory`, 2nd for `ubiquity/devpool-rfc`) hence reduce chances for being rate limited.

xno-miner committed:
Create env variables for devpool/RFC

xno-miner commented:
I added the env variables. For devpool their values should be:
```
DEVPOOL_OWNER_NAME=ubiquity
DEVPOOL_REPO_NAME=devpool-directory
RFC=0
```
and for RFC they should be:
```
DEVPOOL_OWNER_NAME=ubiquity
DEVPOOL_REPO_NAME=devpool-rfc
RFC=1
```

xno-miner commented:
What's the progress of the review?

rndquu reviewed (changes requested):
@xno-miner 

It seems the code hasn't been run because `process.env.RFC` is always evaluating to `true` ([here](https://github.com/ubiquity/devpool-directory/pull/1212/files#diff-7f46e46a8d914909f1e349297b52bd5ecad7f9c4d15c83ac06f75f8981d9ebc5R392), [here](https://github.com/ubiquity/devpool-directory/pull/1212/files#diff-7f46e46a8d914909f1e349297b52bd5ecad7f9c4d15c83ac06f75f8981d9ebc5R412), [here](https://github.com/ubiquity/devpool-directory/pull/1212/files#diff-7f46e46a8d914909f1e349297b52bd5ecad7f9c4d15c83ac06f75f8981d9ebc5R496), etc...). Env variables are always a string so it doesn't matter if we set `process.env.RFC` to `0` or `1` because those are not empty string always evaluated to `true`.

Pls:
1. Fix `process.env.RFC` (cast to number?)
2. Provide QA examples of both devpool issues and RFC issues that clearly shows that the code is working as expected
3. Resolve code conflicts

xno-miner committed:
Merge branch 'development' into development

xno-miner committed:
Create tests for RFCs

xno-miner committed:
Add missing files

xno-miner commented:
>     1. Fix `process.env.RFC` (cast to number?)

I created a variable `IS_RFC` in `github.ts` which checks if the env variable is equal to `"1"`. To make it consistent i also created variables `DEVPOOL_OWNER_NAME` and `DEVPOOL_REPO_NAME` which just load from the env (it's also how it was before i moved them to env)

>     2. Provide QA examples of both devpool issues and RFC issues that clearly shows that the code is working as expected

I have created a test suite for RFC issues, based on the test suite for devpool, and it appears to run successfully
```
------------|---------|----------|---------|---------|-------------------------------------------------------------------------------
File        | % Stmts | % Branch | % Funcs | % Lines | Uncovered Line #s                                                             
------------|---------|----------|---------|---------|-------------------------------------------------------------------------------
All files   |   62.73 |    60.58 |   57.77 |   64.19 |                                                                               
 github.ts  |   64.25 |    61.29 |   60.46 |   66.06 | 62-69,114,183-187,233,250-255,291-345,349-391,419-420,425-437,498,594,629,642 
 twitter.ts |   45.45 |    53.33 |       0 |   45.45 | 12,29-47                                                                      
------------|---------|----------|---------|---------|-------------------------------------------------------------------------------
Test Suites: 1 passed, 1 total
Tests:       43 passed, 43 total
Snapshots:   0 total
Time:        5.518 s
Ran all test suites matching /rfc-issue-handler/i.
Done in 6.00s.
```
I disabled it from being ran automatically for now, beacause it requires a different .env, but once the repository will be duplicated there will be 2 .envs and this won't be an issue. The instructions to run the test suite with the regular .env are in the test suite file (`tests/rfc-issue-handler.test.ts.skip`).

>     3. Resolve code conflicts

Done. Just had to move the statistics logic in `index.ts`.

0x4007 commented on helpers/github.ts:
```diff
@@ -7,6 +7,16 @@ import { writeFile } from "fs/promises";
+// Usually
+export const DEVPOOL_OWNER_NAME = process.env.DEVPOOL_OWNER_NAME;
+export const DEVPOOL_REPO_NAME = process.env.DEVPOOL_REPO_NAME;
+export const IS_RFC = process.env.RFC == "1";
```
I would JSON.parse this environment variable for a boolean. Doesn't make sense to me why you would use a number in place of a boolean.

xno-miner replied:
done

xno-miner committed:
JSON.parse RFC env

rndquu commented:
@xno-miner From [rfc-issue-handler.test.ts.skip](https://github.com/ubiquity/devpool-directory/blob/ed3bbf0825fc689a3f7a4f121615191cdea87b4d/tests/rfc-issue-handler.test.ts.skip#L11-L14):
```
// To Run:
// - Remove .skip from the file name
// - Comment the "Usually" segment in helpers/github.ts and uncomment the "rfc-issue-handler.test.ts" segment
// - Run `yarn test rfc-issue-handler`
```

1. Why `rfc-issue-handler` tests are skipped? 
2. Why do we need to comment some code in `helpers/github.ts` and uncomment some other part in order to run the tests? Can't we mock the environment in the tests?
3. The errors in [this](https://github.com/ubiquity/devpool-directory/actions/runs/10143966190/job/28082954701?pr=1212) CI run are typescript compile errors and are not related to empty env variables.

xno-miner committed:
quotes in ymls

xno-miner committed:
Fix env and enable rfc test

xno-miner committed:
minor cleanup

xno-miner commented:
Ok. I managed to fix this. I just needed to add exclamation marks after env variables to unwrap them. Now there are no compile errors when testing.

I also enabled `rfc-issue-handler` tests, with modified env variables (which now worked because there were no compile errors).

rndquu commented on helpers/github.ts:
```diff
@@ -7,6 +7,10 @@ import { writeFile } from "fs/promises";
 import twitter from "./twitter";
 import { TwitterMap } from "..";
 
+export const DEVPOOL_OWNER_NAME = process.env.DEVPOOL_OWNER_NAME!;
+export const DEVPOOL_REPO_NAME = process.env.DEVPOOL_REPO_NAME!;
+export const IS_RFC = JSON.parse(process.env.RFC!);
```
Non-null assertion operator is considered a bad practice because if the value is `null` then the app fails while the compiler is trying to make sure that even if the value is `null` the app works as expected (or at least throws meaningful errors). The `value!` syntax is not used anywhere in the codebase. Pls refactor accordingly.

0x4007 replied:
I imagine that the linter settings are not in sync with our template here, given that this passed CI. 

Perhaps this is technically our responsibility to handle these sort of problems.

rndquu replied:
> I imagine that the linter settings are not in sync with our template here, given that this passed CI.
> 
> Perhaps this is technically our responsibility to handle these sort of problems.

Seems you've already created a related issue https://github.com/ubiquity/ts-template/issues/54

xno-miner commented:
Closing in favor of #1250
